File: src/test/java/com/conference/website/service/E03_TalkServiceTest.java
package com.conference.website.service;

import com.conference.website.data.builders.*;
import com.conference.website.domain.Speaker;
import com.conference.website.domain.Tag;
import com.conference.website.domain.Talk;
import com.conference.website.domain.TalkLevel;
import com.conference.website.dto.*;
import com.conference.website.repository.SpeakerRepository;
import com.conference.website.repository.TagRepository;
import com.conference.website.repository.TalkRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest
@Transactional
class E03_TalkServiceTest {

    @Autowired
    private SpeakerService speakerService;

    @Autowired
    private SpeakerRepository speakerRepository;

    @Autowired
    private TagRepository tagRepository;

    @Autowired
    private TalkRepository talkRepository;

    @Autowired
    private TalkService talkService;

    @Test
    void shouldCreateMultipleTalksWithEntityBuildersAndTemporaryVariables() {
        //Arrange
        Speaker primarySpeakerTalkOne = SpeakerBuilder.aSpeaker()
                .withName("Ada Lovelace")
                .withEmail("ada@example.com")
                .withBio("Pioneer in computing")
                .build();

        Speaker coSpeakerTalkOne = SpeakerBuilder.from(primarySpeakerTalkOne)
                .withName("Grace Hopper")
                .withEmail("grace@example.com")
                .withBio("COBOL pioneer")
                .build();

        Speaker primarySpeakerTalkTwo = SpeakerBuilder.aSpeaker()
                .withName("Linus Torvalds")
                .withEmail("linus@example.com")
                .withCompany("Kernel Inc")
                .withBio("Created Linux")
                .build();

        Tag kotlinTag = TagBuilder.aTag()
                .withName("kotlin")
                .build();

        Tag testingTag = TagBuilder.aTag()
                .withName("testing")
                .build();

        Tag springTag = TagBuilder.aTag()
                .withName("spring")
                .build();

        Talk talkEntityOne = TalkBuilder.aTalk()
                .withTitle("Kotlin DSL Power")
                .withAbstractText("Scope fixtures without temporary variables")
                .withLevel(TalkLevel.INTERMEDIATE)
                .withDurationMinutes(45)
                .withPrimarySpeaker(primarySpeakerTalkOne)
                .withCoSpeaker(coSpeakerTalkOne)
                .withTag(kotlinTag)
                .withTag(testingTag)
                .build();

        Talk talkEntityTwo = TalkBuilder.aTalk()
                .withTitle("Spring Testing at Scale")
                .withAbstractText("Keep setup readable while growing scenarios")
                .withLevel(TalkLevel.ADVANCED)
                .withDurationMinutes(60)
                .withPrimarySpeaker(primarySpeakerTalkTwo)
                .withTags(List.of(springTag))
                .build();

        TalkGraphPersistence.persistGraph(
                List.of(talkEntityOne, talkEntityTwo),
                speakerRepository,
                tagRepository,
                talkRepository
        );





        

        //Act
        List<TalkDto> createdTalks = talkService.listTalks();

        //Assert
        TalkDto createdTalkOne = createdTalks.getFirst();
        TalkDto createdTalkTwo = createdTalks.getLast();

        assertThat(createdTalks).hasSize(2);
        assertThat(createdTalks)
                .extracting(TalkDto::title)
                .containsExactlyInAnyOrder("Kotlin DSL Power", "Spring Testing at Scale");

        assertThat(createdTalkTwo.primarySpeaker().name())
           .isEqualTo("Ada Lovelace");
        assertThat(createdTalkTwo.coSpeakers())
           .extracting(SpeakerDto::name)
           .containsExactly("Grace Hopper");
        assertThat(createdTalkTwo.tags())
           .extracting(TagDto::name)
           .containsExactlyInAnyOrder("kotlin", "testing");

        assertThat(createdTalkOne.primarySpeaker().name())
           .isEqualTo("Linus Torvalds");
        assertThat(createdTalkOne.tags())
           .extracting(TagDto::name).containsExactly("spring");
    }
}

File: src/test/kotlin/com/conference/website/service/E03_TalkServiceTest.kt
package com.conference.website.service

import com.conference.website.domain.Speaker
import com.conference.website.domain.Tag
import com.conference.website.domain.Talk
import com.conference.website.domain.TalkLevel
import com.conference.website.dto.SpeakerDto
import com.conference.website.dto.TagDto
import com.conference.website.dto.TalkDto
import com.conference.website.repository.SpeakerRepository
import com.conference.website.repository.TagRepository
import com.conference.website.repository.TalkRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@Transactional
class E03_TalkServiceTest {

    @Autowired
    private lateinit var speakerRepository: SpeakerRepository

    @Autowired
    private lateinit var tagRepository: TagRepository

    @Autowired
    private lateinit var talkRepository: TalkRepository

    @Autowired
    private lateinit var talkService: TalkService

    @Test
    fun shouldCreateMultipleTalksWithTemporaryVariables() {
        //Arrange
        val primarySpeakerTalkOne = Speaker(
            "Ada Lovelace",
            "ada@example.com",
            "Analytical Engines",
            "Pioneer in computing"
        )

        val coSpeakerTalkOne = Speaker(
            "Grace Hopper",
            "grace@example.com",
            primarySpeakerTalkOne.company, //<- copy a field by hand
            "COBOL pioneer"
        )

        val primarySpeakerTalkTwo = Speaker(
            "Linus Torvalds",
            "linus@example.com",
            "Kernel Inc",
            "Created Linux"
        )

        val kotlinTag = Tag("kotlin")

        val testingTag = Tag("testing")

        val springTag = Tag("spring")

        val talkEntityOne = Talk(
            "Kotlin DSL Power",
            "Scope fixtures without temporary variables",
            TalkLevel.INTERMEDIATE,
            45,
            primarySpeakerTalkOne
        )
        talkEntityOne.coSpeakers = linkedSetOf(coSpeakerTalkOne)
        talkEntityOne.tags = linkedSetOf(kotlinTag, testingTag)

        val talkEntityTwo = Talk(
            "Spring Testing at Scale",
            "Keep setup readable while growing scenarios",
            TalkLevel.ADVANCED,
            60,
            primarySpeakerTalkTwo
        )
        talkEntityTwo.tags = linkedSetOf(springTag)

        //every temporary variable has to be listed again, in the right order,
        //and it is on us to remember that speakers and tags come before talks
        speakerRepository.saveAll(
            listOf(primarySpeakerTalkOne, coSpeakerTalkOne, primarySpeakerTalkTwo)
        )
        tagRepository.saveAll(listOf(kotlinTag, testingTag, springTag))
        talkRepository.saveAll(listOf(talkEntityOne, talkEntityTwo))
        talkRepository.flush()


        //Act
        val createdTalks: List<TalkDto> = talkService.listTalks()



        //Assert
        val createdTalkOne = createdTalks.first()
        val createdTalkTwo = createdTalks.last()

        assertThat(createdTalks).hasSize(2)
        assertThat(createdTalks)
            .extracting<String>(TalkDto::title)
            .containsExactlyInAnyOrder("Kotlin DSL Power", "Spring Testing at Scale")

        assertThat(createdTalkTwo.primarySpeaker.name)
            .isEqualTo("Ada Lovelace")
        assertThat(createdTalkTwo.coSpeakers)
            .extracting<String>(SpeakerDto::name)
            .containsExactly("Grace Hopper")
        assertThat(createdTalkTwo.tags)
            .extracting<String>(TagDto::name)
            .containsExactlyInAnyOrder("kotlin", "testing")

        assertThat(createdTalkOne.primarySpeaker.name)
            .isEqualTo("Linus Torvalds")
        assertThat(createdTalkOne.tags)
            .extracting<String>(TagDto::name).containsExactly("spring")
    }
}
